Migrate to aws-sdk-go-v2

zeleena opened the pull request:
Big picture Lyft-specific goals: Fedora CoreOS pulls ignition configs from S3. To make FCOS backplane rollout more robust to S3 failure, we'd like to use Multi-Region Access Points for region fallback, but this feature is only available in aws-sdk-go-v2.

The first step is to migrate ignition's aws-sdk-go v1 to aws-sdk-go-v2 (**this PR completes this first step**).
The second step is to make S3 URLs with access points compatible with RFC3986 (due to an invalid authority portion) by implementing a URI change to `scheme:path`. Then S3 URLs will be a valid opaque URI that net/curl can parse and represent.

The bulk of this change is in the first commit with changes to files `internal/providers/aws/aws.go` and `internal/resource/url.go`. I used the aws-sdk-go-v2 migration guide (https://aws.github.io/aws-sdk-go-v2/docs/migrating/) and documentation (https://pkg.go.dev/github.com/aws/aws-sdk-go-v2). The dependency changes are a product of running `make vendor`.

Following [development.md](https://github.com/coreos/ignition/blob/main/docs/development.md), I tested on a Fedora node by running the following:
 ```
# ./build_blackbox_tests 
Using version from git...
Building ignition...
Building ignition-validate...
Compiling tests...
Building tests/stubs/groupdel-stub
Building tests/stubs/useradd-stub
Building tests/stubs/userdel-stub
Building tests/stubs/usermod-stub
Success
# git describe --dirty --always
v2.12.0-12-g227a7e7c
# sh -c 'PATH=$PWD/bin/amd64:$PATH ./tests.test'
PASS
```

zeleena committed:
Migrate to aws-sdk-go-v2 with changes in aws.go and resource/url.go

zeleena committed:
Migrate to aws-sdk-go-v2 with changes in aws.go and resource/url.go

zeleena committed:
Ran 'make vendor' to remove v1 and get v2 modules

zeleena committed:
Ran 'make vendor' to remove v1 and get v2 modules

zeleena commented:
meow #compute @paulnivin @ashleykasim @dghubble

zeleena committed:
Use context.Background and fix the way downloader uses anonymous credentials

zeleena committed:
Use context.Background and fix the way downloader uses anonymous credentials

zeleena committed:
Run 'make vendor' for latest change

zeleena committed:
Run 'make vendor' for latest change

paulnivin commented on internal/resource/url.go:
```diff
@@ -458,25 +469,28 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
+			_, err = downloader.Download(ctx, dest, input,
+				manager.WithDownloaderClientOptions(func(o *s3.Options) {
+					o.Credentials = aws.AnonymousCredentials{}
+					o.HTTPClient = httpClient
```
shouldn't need to override HTTP client again as s3Client already has it overridden

zeleena replied:
done, removed line

paulnivin commented on internal/resource/url.go:
```diff
@@ -396,31 +395,43 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
 		regionHint = f.S3RegionHint
 	}
-	region, err := s3manager.GetBucketRegion(ctx, sess, u.Host, regionHint)
+	region, err := manager.GetBucketRegion(ctx, s3.NewFromConfig(cfg), u.Host)
```
s3 client should use regionHint as it's region

zeleena replied:
done, kept s3Client initialized from NewFromConfig here and in the helper. I see you only initialized it once and made it a field for fetcher, let me know if I should do that as well

paulnivin commented on internal/resource/url.go:
```diff
@@ -396,31 +395,43 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
+			f.Logger.Warning("couldn't determine the region for bucket %q: %v, trying again with region hint", u.Host, err)
+		}
+		// Use the region hint to determine the partition and region this bucket is in
+		region, err = manager.GetBucketRegion(ctx, s3.NewFromConfig(cfg), u.Host, func(o *s3.Options) {
```
NewFromConfig(....WithRegion(regionHint) for the block above should mean that this block can be removed

zeleena replied:
done, removed this block

paulnivin commented on internal/resource/url.go:
```diff
@@ -396,31 +395,43 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
 	}
 
-	sess.Config.Region = aws.String(region)
+	cfg.Region = *aws.String(region)
```
Region is just a normal string here

zeleena replied:
done, fixed in aws.go around regionHint too

paulnivin commented on internal/providers/aws/aws.go:
```diff
@@ -58,15 +59,15 @@ func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
 		return resource.Fetcher{}, err
 	}
-	sess.Config.Credentials = ec2rolecreds.NewCredentials(sess)
+	cfg.Credentials = aws.NewCredentialsCache(ec2rolecreds.New())
```
this line can be dropped. LoadDefaultConfig when run on an EC2 instance will use the EC2 creds

zeleena replied:
cool! removed line and updated vendoring

zeleena committed:
Use regionHint when getting bucket region and simplify

zeleena committed:
Use regionHint when getting bucket region and simplify

zeleena committed:
Reran 'make vendor' for last change

zeleena committed:
Reran 'make vendor' for last change

sohankunkerkar committed:
internal/*: change the location of Ignition report

Fixes https://github.com/coreos/fedora-coreos-tracker/issues/977

This change helps to address the problem of displaying a warning
about running Ignition twice while reprovisioning a bare metal
CoreOS instance with a persistent /var partition.

jlebon committed:
internal/exec/util: rename FindFirstMissingDirForFile and tweak docs

It actually returns a path, not a component. Also, it can return the
path itself if just the final component is missing or if it already
exists. Let's tweak the docstring and use the term "component" in its
name to be clearer.

travier committed:
docs: Do not convert -- & --- to en/em-dash

'--' is frequently used for command line options and was thus
incorrectly rendered as a special en-dash symbol.

travier committed:
docs: Remove default layout from front matter

No need to specify it as it is the default.

jlebon committed:
Merge pull request #1270 from travier/docfixes

jlebon committed:
Merge pull request #1268 from jlebon/pr/tweak-first-missing-name

bgilbert committed:
docs/supported-platforms: add platform IDs

bgilbert committed:
docs/supported-platforms: sort by platform ID

bgilbert committed:
docs/supported-platforms: update platform names and URLs

bgilbert committed:
docs/supported-platforms: add missing platforms

Omit VirtualBox for now, since our current userdata mechanism there isn't
very usable.

Fixes https://github.com/coreos/ignition/issues/1222.

bgilbert committed:
test: ensure all platforms are documented

Skip VirtualBox for now, since our current userdata mechanism there isn't
very usable.  Omit "file" since it's a test platform.

bgilbert committed:
docs/supported-platforms: drop reference to platform-specific agents

This is a holdover from Container Linux.

bgilbert committed:
docs/supported-platforms: switch to Afterburn docs URL

sohankunkerkar committed:
Merge pull request #1267 from sohankunkerkar/fix-ignition-prev-run

internal/*: change the location of Ignition report

jlebon committed:
stages/files: make variable name follow Go convention

My editor told me "don't use underscores in Go names".

jlebon committed:
stages/files: rename `relabelDirsForFile` and add docstring

Follow-up to #1268.

bgilbert committed:
Merge pull request #1271 from bgilbert/platforms

docs/supported-platforms: update for current platform list

jlebon committed:
Merge pull request #1272 from jlebon/pr/more-name-tweaking

Commit:
stages/filesystems: use mkfs.fat instead of mkfs.vfat

Ignition should use mkfs.fat instead of mkfs.vfat because mkfs.vfat is
symlink to mkfs.fat and may be absent in some distributions.
Dosfstools always creates mkfs.fat, but mkfs.vfat is optional
(https://github.com/dosfstools/dosfstools/blob/master/src/Makefile.am)

bgilbert committed:
Merge pull request #1273 from keremet/main

stages/filesystems: use mkfs.fat instead of mkfs.vfat

zeleena commented:
/rebase

zeleena committed:
Merge branch 'migrate-to-aws-sdk-go-v2' of github.com:lyft/ignition into migrate-to-aws-sdk-go-v2

dghubble reviewed (commented):
noop review, bc I don't seem to have permission to remove myself as a reviewer
